merino-suggest/src/keyword_filter.rs

```diff
@@ -0,0 +1,223 @@
+//! A suggestion provider that filters suggestions from a subprovider.
+
+use std::collections::HashMap;
+
+use crate::{SetupError, SuggestionProvider, SuggestionResponse};
+use anyhow::anyhow;
+use async_trait::async_trait;
+use cadence::{Counted, StatsdClient};
+use regex::Regex;
+
+/// A combinator provider that filters the results from the wrapped provider
+/// using a blocklist from the settings.
+pub struct KeywordFilterProvider {
+    /// A map linking the filter identifiers to their compiled regexes.
+    blocklist: HashMap<String, regex::Regex>,
+
+    /// The provider to pull suggestions from.
+    inner: Box<dyn SuggestionProvider>,
+}
+
+impl KeywordFilterProvider {
+    /// Construct a new, boxed filter provider.
+    pub fn new_boxed(
+        blocklist: HashMap<String, String>,
+        inner: Box<dyn SuggestionProvider>,
+    ) -> Result<Box<Self>, SetupError> {
+        // Compile the provided blocklist regexes just once.
+        let mut compiled_blocklist: HashMap<String, Regex> = HashMap::new();
+
+        for filter in &blocklist {
+            if let Ok(r) = Regex::new(filter.1) {
```

Comment:
I think making filters case sensitive is pretty jarring to those that would be configuring the system. I think making the regex case insensitive would be good:

```suggestion
            if let Ok(r) = RegexBuilder::new(filter.1).case_insensitive(true).build() {
```